### PR TITLE
identity propagation in ssh secrets engine #7547

### PR DESCRIPTION
**Pull request by fredrikhgrelland**
To test you may run:
```
pkill -f -9 vault
make dev-ui
export VAULT_DEV_ROOT_TOKEN_ID=test
export VAULT_TOKEN=test
vault server -dev -dev-no-store-token&
sleep 5
export VAULT_ADDR=http://127.0.0.1:8200
vault token lookup
cat <<EOF > user-policy.hcl
# Configure auth methods
path "/*" {
  capabilities = [ "create", "update", "read", "delete", "list", "sudo" ]
}
EOF
vault policy write user user-policy.hcl
vault auth enable userpass
vault write auth/userpass/users/ubuntu password=test policies=user
vault secrets enable -path=ssh ssh
vault write ssh/config/ca generate_signing_key=true
cat <<EOF > signer-clientrole.hcl
{
    "allow_user_certificates": true,
    "allowed_users": "root,{{identity.entity.aliases.$(vault auth list -format=json | jq -r '.["userpass/"].accessor').name}}",
    "default_user": "",
    "allow_user_key_ids": "false",
    "default_extensions": [
        {
          "permit-pty": ""
        }
    ],
    "key_type": "ca",
    "ttl": "60m0s"
}
EOF
vault write ssh/roles/clientrole @signer-clientrole.hcl
vault token lookup
vault write \
    -field=signed_key \
    ssh/sign/clientrole \
    valid_principals="root" \
    public_key=@$HOME/.ssh/id_rsa.pub \
    > $HOME/.ssh/cert-signed.pub
ssh-keygen -L -f $HOME/.ssh/cert-signed.pub
export VAULT_TOKEN=$(vault login -token-only -method=userpass username=ubuntu password=test)
vault token lookup
vault write \
    -field=signed_key \
    ssh/sign/clientrole \
    valid_principals="ubuntu" \
    public_key=@$HOME/.ssh/id_rsa.pub \
    > $HOME/.ssh/cert-signed.pub
ssh-keygen -L -f $HOME/.ssh/cert-signed.pub

```

### CODE REVIEW

**Comment by jefferai** [comment]
I don't think we can accept this as is. DisplayName is often not set by auth methods and can sometimes be set by the user. We don't rely on it anywhere else for these reasons and consider it more or less unused and untrustworthy.

The better approach is to grant full access to Identity information from plugins via enhancing sysview. Right now if you know an entity ID you can get entityinfo but in most cases the entity ID isn't known. The entity ID needs to be sent in requests across the plugin barrier and we need to enhance the info in Entity, or we need to add a different call that can return group information as well.

Then the question is how to control whether templating it allowed in plugin fields but that's likely a per-plugin determination.

**Comment by fredrikhgrelland** [comment]
@jefferai Thank you for your swift reply.
I totally understand why you cant merge as is. I have tried to figure out a way to lookup entity but i came up short.
> Right now if you know an entity ID you can get entityinfo
Can you please point me in the right direction for this? req.EntityID contains the entity_id in this context, so if there is a way of getting to more entity information I am happy to do more work on this matter.

**Comment by jefferai** [comment]
@briankassouf ^ We've wanted this a while, maybe someone can help sort out the specifics and guide this along?

**Comment by fredrikhgrelland** [comment]
@jefferai I was doing some more investigation and found your comments in  #4681 and that led me to the implemetation proposed here: https://github.com/paxosglobal/vault-plugin-secrets-kubernetes/blob/bbaf746dca587b4a7506d4ad72795435726a9e76/pkg/backend/path_token.go
This has been implemented, with minor changes and pushed. Will this do the trick?

**Comment by fredrikhgrelland** [comment]
@jefferai Applogies for the bump. I made some changes to the code, using entities/aliases and falling back on display name if it fails. I also looked at the linked issue where using templating is proposed, but for this use case I am fine with a simple switch and do not need full templating support. Please let me know if there is anything I can do to help this feature be realized.

**Comment by prbinu** [comment]
Both in Vault SSH and PKI/X509 engines doesn't have a way (hope I'm wrong) to include the current user id. As an enterprise Vault user, we are finding extremely difficult to identify the actual users (not the roles) on our servers, be it SSH or web services using X.509 client certs. The display_name cannot be trusted [1]. There should be some easy way to address this issue. If there is a way to do this, please share the steps on how to enable it.

[1] https://www.vaultproject.io/api/secret/ssh/index.html
key_id_format (string: "") – When supplied, this value specifies a custom format for the key id of a signed certificate. The following variables are available for use: '{{token_display_name}}' - The display name of the token used to make the request. '{{role_name}}' - The name of the role signing the request. '{{public_key_hash}}' - A SHA256 checksum of the public key that is being signed. e.g. "custom-keyid-{{**token_display_name**}}"

**Comment by selurvedu** [comment]
+1. My use case is exactly the same as @prbinu's.

**Comment by fredrikhgrelland** [comment]
> @jefferai I was doing some more investigation and found your comments in #4681 and that led me to the implemetation proposed here: https://github.com/paxosglobal/vault-plugin-secrets-kubernetes/blob/bbaf746dca587b4a7506d4ad72795435726a9e76/pkg/backend/path_token.go
> This has been implemented, with minor changes and pushed. Will this do the trick?

> ...I made some changes to the code, using entities/aliases and falling back on display name if it fails. ... Please let me know if there is anything I can do to help this feature be realized.

> @briankassouf ^ We've wanted this a while, maybe someone can help sort out the specifics and guide this along?

@briankassouf May I solicit a response on this PR? As others have pointed out, not being able to identify an individual authenticating using the SSH backend is major blocker in any setting that require auditing of user access.
This PR introduces a very simple addition that will allow us to trust the identity backends as the authoritative source. In our case we want to use sssd in rhel to sync Active Directory user/group mappings, but authenticate based on SSH cert from vault. The user logging on to a server using Vault SSH will, with this change, forcibly be mapped to the username as it is exposed by the Active Directory LDAP backend in Vault.

File: builtin/logical/ssh/path_sign.go [hunk 1]
```diff
@@ -228,6 +228,29 @@ func (b *backend) calculateValidPrincipals(data *framework.FieldData, defaultPri
...
+		if principalsAllowedByRole == "?" {
+			// Use the authenticated users entity alias as username.
+			var username string
+			if req.EntityID != "" {
+				entity, err := b.System().EntityInfo(req.EntityID)
+				if err != nil {
+					return nil, err
+				}
+				if len(entity.Aliases) > 0 {
+					// Take the first alias available for the username.
```
Comment (ianferguson) [inline builtin/logical/ssh/path_sign.go]: Is it worth considering setting the entity name to the entity alias's name when a brand new entity is created for an alias?

i.e. in this code: https://github.com/hashicorp/vault/blob/c27a1c1840a4f73674c5a511b7a2ed5a44cc73bb/vault/identity_store.go#L530-L531, if entity.Name were set to the entity alias name after the entity struct was initialized, then the code here (and other similar use cases) would not need to guess/use their own heuristic to find a useful principal name from the entity's aliases.

**Comment by briankassouf** [comment]
@fredrikhgrelland Thanks for submitting this PR, and apologies for the delay. This is something we'd like to get merged in, but needs a few small changes. Instead of using a `?` to indicate using the logged in user it would be more flexible to use our [identity templating framework](https://www.vaultproject.io/docs/concepts/policies.html#templated-policies). This would enable the user to determine which identity value they want to be represented as the username, rather than having hard coded logic. #8088 was just opened to expose this templating logic to secret engines. Once that PR is merged it would need to be used by the SSH secret engine. 

Likely we'd want to add an extra boolean role argument which specifies if the provided username is an identity template or not. If so, it should be populated by calling `framework. PopulateIdentityTemplate()`.

**Comment by prbinu** [comment]
Thank you @fredrikhgrelland and @briankassouf.
It would be of great help if you can extend this feature to PKI X509 certificates as well.

**Comment by selurvedu** [comment]
#8088 seems to be merged now.

File: builtin/logical/ssh/path_sign.go (hunk 1 above)
Comment (fredrikhgrelland) [reply builtin/logical/ssh/path_sign.go]: Using identity template.

**Comment by fredrikhgrelland** [comment]
Thank you for your reply @briankassouf.
I rewrote the patch to use the identity templating framework. It is a pretty slick feature!
> Likely we'd want to add an extra boolean role argument which specifies if the provided username is an identity template or not.
With the implementation in this patch it would not be necessary to do so. It uses a simple regex to resolve both templates and static principals.
Please let me know if I am on the right path here.

**Comment by jefferai** [comment]
> With the implementation in this patch it would not be necessary to do so. It uses a simple regex to resolve both templates and static principals.

The point of this isn't to figure out whether there are templating markers, but whether they should be allowed. It's not clear that we should always allow templating. (I'm not sure it's clear we shouldn't, either.)

**Comment by fredrikhgrelland** [comment]
> > With the implementation in this patch it would not be necessary to do so. It uses a simple regex to resolve both templates and static principals.
> 
> The point of this isn't to figure out whether there are templating markers, but whether they should be allowed. It's not clear that we should always allow templating. (I'm not sure it's clear we shouldn't, either.)

Ok. Would you like me to take a stab at it?
My personal opinion: I do not see why templating should or should not be "allowed" by using a flag. This would be used is in the context of configuring a role. The flag should mean something for the administrator, or specifically open a code path you do not normally want executed.

**Comment by fredrikhgrelland** [comment]
@jefferai  421707a6ecab1d4c3b8db5213633dc01f3a153e9 adds a boolean flag to enable templating for allowed_users. If that helps?

**Comment by fredrikhgrelland** [comment]
@jefferai or @briankassouf I appreciate that you are working on external PRs on a best-effort basis, but a couple of weeks has passed, and I am wondering if I am wasting my time and yours on this effort.

**Pull request merged**